### PR TITLE
fix turning function items into closure trait objects

### LINKED ISSUE #111: fn item as closure trait object "index out of bounds"
**Source**

```rust
fn foo() {}

fn main() {
    let f: &Fn() = &foo;
    f();
}
```

**Output and error**

```
TRACE:miri::step 0  _4 = <std::ops::Fn() as std::ops::Fn<()>>::call(_5, _6) -> bb1
TRACE:miri::eval_context 0  frame[0] _4: Undef
TRACE:miri::eval_context 0  frame[0] _5: (Ptr(Pointer { alloc_id: AllocId(0), offset: 0 }), Ptr(Pointer { alloc_id: AllocId(3), offset: 0 }))
TRACE:miri::memory 0  Alloc 3:    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 (48 bytes) (immutable)
TRACE:miri::memory 0                                                                                      └─────────(2)─────────┘ └─────────(2)─────────┘ └─────────(2)─────────┘                                    
TRACE:miri::memory 0  Alloc 2:     function pointer
TRACE:miri::eval_context 0  frame[0] _6:
DEBUG:miri::memory: reading fn ptr: 2
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', src/terminator/mod.rs:632
```

### LINKED ISSUE #112: fn pointer as closure trait object "bad VtableFnPointer fn_ty: fn()"
**Source**

```rust
fn foo() {}

fn main() {
    let f: &Fn() = &(foo as fn());
    f();
}
```

**Output and error**

```
TRACE:miri::step 0  _1 = _2 as &std::ops::Fn() (Unsize)
TRACE:miri::eval_context 0  frame[0] _1: Undef
TRACE:miri::eval_context 0  frame[0] _2: Ptr(Pointer { alloc_id: AllocId(3), offset: 0 })
TRACE:miri::memory 0  Alloc 3:    00 00 00 00 00 00 00 00 (8 bytes) (immutable)
TRACE:miri::memory 0              └─────────(2)─────────┘
TRACE:miri::memory 0  Alloc 2:     function pointer
DEBUG:miri::vtable: get_vtable(trait_ref=Binder(<fn() as std::ops::Fn<()>>))
error: internal compiler error: src/vtable.rs:63: bad VtableFnPointer fn_ty: fn()       
```

### LINKED ISSUE #114: multi-argument closure virtual calls are broken
**Source**

From [this rustc test](https://github.com/rust-lang/rust/blob/master/src/test/run-pass/overloaded-calls-object-two-args.rs).

```rust
fn foo(f: &mut FnMut(isize, isize) -> isize) -> isize {
    f(1, 2)
}

fn main() {
    let z = foo(&mut |x, y| x * 10 + y);
    assert_eq!(z, 12);
}
```

**Output and error**

```
TRACE:miri::step:   _0 = <std::ops::FnMut(isize, isize) -> isize as std::ops::FnMut<(isize, isize)>>::call_mut(_3, _4) -> bb1
TRACE:miri::eval_context:   _1 (1 frames up): Undef
TRACE:miri::eval_context:   _3: (Ptr(Pointer { alloc_id: AllocId(0), offset: 0 }), Ptr(Pointer { alloc_id: AllocId(3), offset: 0 }))
TRACE:miri::memory:   Alloc 3:    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 (40 bytes) (immutable)
TRACE:miri::memory:                                                                                       └─────────(2)─────────┘ └─────────(2)─────────┘ 
TRACE:miri::memory:   Alloc 2:     function pointer: main::{{closure}}: extern "rust-call" fn((isize, isize)) -> isize
TRACE:miri::eval_context:   _4:
TRACE:miri::memory:   Alloc 4:    01 00 00 00 00 00 00 00 02 00 00 00 00 00 00 00 (16 bytes)
DEBUG:miri::memory: reading fn ptr: 2
TRACE:miri::eval_context:   load mir DefId { krate: CrateNum(0), node: DefIndex(7) => overloaded_calls_object_two_args/4089d7c8b778d88cec885baf7b69e6df-exe::main[0]::{{closure}}[0] }
TRACE:miri::eval_context:    _1: Undef
TRACE:miri::eval_context:    _2: Undef
TRACE:miri::step:    // bb0
TRACE:miri::step:    StorageLive(_4)
TRACE:miri::step:    _4 = _2
TRACE:miri::eval_context:    _4: Undef
TRACE:miri::eval_context:    _2:
TRACE:miri::memory:    Alloc 5:    01 00 00 00 00 00 00 00 02 00 00 00 00 00 00 00 (16 bytes)
TRACE:miri::eval_context:    _4: Bytes(1)
TRACE:miri::step:    StorageLive(_5)
TRACE:miri::step:    _5 = _3
TRACE:miri::eval_context:    _5: Undef
TRACE:miri::eval_context:    _3: Undef
TRACE:miri::eval_context:    _5: Undef
TRACE:miri::step:    StorageLive(_6)
TRACE:miri::step:    StorageLive(_7)
TRACE:miri::step:    _7 = _4
TRACE:miri::eval_context:    _7: Undef
TRACE:miri::eval_context:    _4: Bytes(1)
TRACE:miri::eval_context:    _7: Bytes(1)
TRACE:miri::step:    _8 = CheckedMul(_7, const 10isize)
TRACE:miri::eval_context:    _8: Undef
TRACE:miri::eval_context:    _7: Bytes(1)
TRACE:miri::eval_context:    _8: (Bytes(10), Bytes(0))
TRACE:miri::step:    assert(!(_8.1: bool), "attempt to multiply with overflow") -> bb1
TRACE:miri::step:    // bb1
TRACE:miri::step:    _6 = (_8.0: isize)
TRACE:miri::eval_context:    _6: Undef
TRACE:miri::eval_context:    _6: Bytes(10)
TRACE:miri::step:    StorageLive(_9)
TRACE:miri::step:    _9 = _5
TRACE:miri::eval_context:    _9: Undef
TRACE:miri::eval_context:    _5: Undef
TRACE:miri::eval_context:    _9: Undef
TRACE:miri::step:    _10 = CheckedAdd(_6, _9)
TRACE:miri::eval_context:    _10: Undef
TRACE:miri::eval_context:    _6: Bytes(10)
TRACE:miri::eval_context:    _9: Undef
error: attempted to read undefined bytes
  --> ../rust/master/src/test/run-pass/overloaded-calls-object-two-args.rs:20:29
   |
20 |     let z = foo(&mut |x, y| x * 10 + y);
   |                             ^^^^^^^^^^
```

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #111
fixes #112 
fixes #114

### CODE REVIEW

File: src/terminator/mod.rs [hunk 1]
```diff
@@ -313,6 +313,10 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
                 )?;
 
                 let arg_locals = self.frame().mir.args_iter();
+                // FIXME: impl ExactSizeIterator and use args_locals.len()
```
Comment (solson) [inline src/terminator/mod.rs]: Use `self.frame().mir.arg_count`, it's public.

File: src/terminator/mod.rs [hunk 2]
```diff
@@ -313,6 +313,10 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
+                // FIXME: last-use-in-cap-clause works by chance, insert some arguments and it will fail
+                // we currently write the first argument (unit) to the return field (unit)
```
Comment (solson) [inline src/terminator/mod.rs]: Are you sure we do this? I don't see anything writing to the return lvalue. `mir.args_iter()` doesn't include the return lvalue.

**Comment by solson** [comment]
I wonder if this fixes #114, too.

**Comment by solson** [comment]
Maybe not, but it's a related problem. #114 is having a tuple/bare arguments mismatch.

File: src/terminator/mod.rs (hunk 2 above)
Comment (oli-obk) [inline src/terminator/mod.rs]: I am triggering the assertion in that test, but didn't want to fix it in this PR.

**Comment by oli-obk** [comment]
#114 sounds like it would trigger the mentioned assertion, too

**Comment by oli-obk** [comment]
I wonder if we should stick some extra knowledge into the vtable when creating it, since that is the place we have all the correct info available. Then, when using the vtable, we just trust that info.

**Comment by oli-obk** [comment]
this PR introduces new errors, let's hold off on it and fix all these correctly...


      1 thread 'main' panicked at 'assertion failed: `(left == right)` (left: `[&isize]`, right: `[&isize]`)', src/terminator/mod.rs:643
      1 thread 'main' panicked at 'assertion failed: `(left == right)` (left: `[&Identity, X]`, right: `[i32]`)', src/terminator/mod.rs:643
      1 thread 'main' panicked at 'assertion failed: `(left == right)` (left: `[&Identity, (&T,)]`, right: `[&i32]`)', src/terminator/mod.rs:643

**Comment by solson** [comment]
@oli-obk Instead of in the vtable memory, can you add extra details to the `memory::FunctionDefinition` to handle these weird cases?

**Comment by oli-obk** [comment]
ok, now we have nice error messages even if someone crazy uses unsafe code to extract the drop glue pointer from a vtable and call it as a function.

This means that it's also not possible anymore to create a vtable on the fly and have it "just work". If you set the drop glue function to a real function, miri will abort. If you set `Fn` trait function to normal function, miri will abort. And so on...

**Comment by oli-obk** [comment]
We could even go as far as adding an error variant for using mutable memory as a vtable. Or anything we haven't marked as `vtable` as a vtable.

File: src/terminator/mod.rs [hunk 3]
```diff
@@ -314,9 +314,8 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
 
                 let arg_locals = self.frame().mir.args_iter();
                 // FIXME: impl ExactSizeIterator and use args_locals.len()
-                // FIXME: last-use-in-cap-clause works by chance, insert some arguments and it will fail
-                // we currently write the first argument (unit) to the return field (unit)
-                //assert_eq!(self.frame().mir.args_iter().count(), args.len());
+                assert_eq!(arg_locals.size_hint().0, args.len());
```
Comment (solson) [inline src/terminator/mod.rs]: Why not use `self.frame().mir.arg_count` instead of this size hint?

Comment (solson) [inline src/terminator/mod.rs]: Oops, you just made this change already.

**Pull request merged**